integration/src/main/java/scripts/testscripts/ListWorkspaces.java

```diff
@@ -64,30 +54,40 @@ protected void doCleanup(List<TestUserSpecification> testUsers, WorkspaceApi wor
   }
 
   @Override
-  protected void doUserJourney(TestUserSpecification testUser, WorkspaceApi workspaceApi)
+  protected void doUserJourney(TestUserSpecification testUser, WorkspaceApi firstUserApi)
       throws Exception {
-    // First, check that listing 3 workspaces will return the 3 we created.
-    WorkspaceDescriptionList workspaceList = workspaceApi.listWorkspaces(/*offset=*/0, /*limit=*/3);
+    // For the sake of listing "all" of a user's workspaces we must have a maximum upper limit. Test
+    // users are expected to never have permanent workspaces, so this should be more than sufficient.
+    int MAX_USER_WORKSPACES = 10000;
+
+    // While test users should not have any permanent workspaces, they may have some entries from
+    // previous failures or other tests running in parallel. We therefore cannot assume they only
+    // have 3 workspaces.
+    WorkspaceDescriptionList workspaceList = firstUserApi.listWorkspaces(/*offset=*/0, /*limit=*/
+        MAX_USER_WORKSPACES);
     List<UUID> workspaceIdList = workspaceList.getWorkspaces().stream().map(WorkspaceDescription::getId).collect(
         Collectors.toList());
-    assertThat(workspaceIdList, containsInAnyOrder(equalTo(getWorkspaceId()), equalTo(workspaceId2), equalTo(workspaceId3)));
+    assertThat(workspaceIdList, hasItems(getWorkspaceId(), workspaceId2, workspaceId3));
 
-    // Next, call the endpoint three times with limit 1 and different offsets.
+    // Next, cover the same set of workspaces across 3 pages.
+    int pageSize = MAX_USER_WORKSPACES / 3;
     List<WorkspaceDescription> callResults = new ArrayList<>();
-    callResults.addAll(workspaceApi.listWorkspaces(/*offset=*/0, /*limit=*/1).getWorkspaces());
-    callResults.addAll(workspaceApi.listWorkspaces(/*offset=*/1, /*limit=*/1).getWorkspaces());
-    callResults.addAll(workspaceApi.listWorkspaces(/*offset=*/2, /*limit=*/1).getWorkspaces());
+    callResults.addAll(firstUserApi.listWorkspaces(/*offset=*/0, /*limit=*/pageSize).getWorkspaces());
+    callResults.addAll(firstUserApi.listWorkspaces(/*offset=*/pageSize, /*limit=*/pageSize).getWorkspaces());
+    // pageSize may not be divisible by 3, so cover all remaining workspaces here instead.
+    callResults.addAll(firstUserApi.listWorkspaces(/*offset=*/pageSize*2, /*limit=*/(
+        MAX_USER_WORKSPACES - 2*pageSize)).getWorkspaces());
     List<UUID> callResultIdList = workspaceList.getWorkspaces().stream().map(WorkspaceDescription::getId).collect(
```

Comment:
Should the value of `callResultIdList` be fetched from `callResults` instead of `workspaceList.getWorkspaces()…`?

Reply:
good catch!